File: orchestrator/ratings.py
"""
Agent Ratings System.

After each task execution, peer agents submit a 0–100 percentage score for
the performing agent.  The system maintains a running weighted average
(tally) for every agent and exposes a competitive leaderboard so agents
can race to stay at the top.
"""

import logging
from collections import defaultdict
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)

_MAX_HISTORY = 500  # rating entries kept per agent


class RatingEntry:
    """A single peer rating submitted after one task."""

    __slots__ = ("agent_id", "rated_by", "task", "score", "comment", "timestamp")

    def __init__(
        self,
        agent_id: str,
        rated_by: str,
        task: str,
        score: float,
        comment: str = "",
    ) -> None:
        if not (0.0 <= score <= 100.0):
            raise ValueError(f"Score must be 0–100, got {score}")
        self.agent_id = agent_id
        self.rated_by = rated_by
        self.task = task
        self.score = round(score, 2)
        self.comment = comment
        self.timestamp = datetime.now(timezone.utc).isoformat()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "agent_id": self.agent_id,
            "rated_by": self.rated_by,
            "task": self.task,
            "score": self.score,
            "comment": self.comment,
            "timestamp": self.timestamp,
        }


class AgentRatingTally:
    """Running tally of ratings for one agent."""

    def __init__(self, agent_id: str, agent_type: str) -> None:
        self.agent_id = agent_id
        self.agent_type = agent_type
        self._history: List[RatingEntry] = []

    # ------------------------------------------------------------------

    def add(self, entry: RatingEntry) -> None:
        self._history.append(entry)
        if len(self._history) > _MAX_HISTORY:
            self._history.pop(0)

    @property
    def total_ratings(self) -> int:
        return len(self._history)

    @property
    def average_score(self) -> float:
        if not self._history:
            return 0.0
        return round(sum(e.score for e in self._history) / len(self._history), 2)

    @property
    def last_score(self) -> Optional[float]:
        return self._history[-1].score if self._history else None

    @property
    def trend(self) -> str:
        """Compare last 5 ratings to the 5 before that."""
        if len(self._history) < 2:
            return "neutral"
        recent = [e.score for e in self._history[-5:]]
        older = [e.score for e in self._history[-10:-5]] or recent
        delta = sum(recent) / len(recent) - sum(older) / len(older)
        if delta > 2:
            return "up"
        if delta < -2:
            return "down"
        return "neutral"

    def get_history(self, limit: int = 20) -> List[Dict[str, Any]]:
        return [e.to_dict() for e in self._history[-limit:]]

    def to_dict(self) -> Dict[str, Any]:
        return {
            "agent_id": self.agent_id,
            "agent_type": self.agent_type,
            "average_score": self.average_score,
            "last_score": self.last_score,
            "total_ratings": self.total_ratings,
            "trend": self.trend,
        }


class RatingSystem:
    """
    Singleton-style ratings registry.

    Usage inside the orchestrator::

        ratings = RatingSystem()

        # After agent X finishes a task, all other agents rate it:
        ratings.submit_peer_ratings(
            performing_agent_id="agent-x",
            performing_agent_type="ai_agent",
            task="auto_optimise",
            peer_agents=[("agent-y", "frequency_agent"), ...],
        )

        leaderboard = ratings.get_leaderboard()
    """

    def __init__(self) -> None:
        self._tallies: Dict[str, AgentRatingTally] = {}

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def ensure_agent(self, agent_id: str, agent_type: str) -> AgentRatingTally:
        if agent_id not in self._tallies:
            self._tallies[agent_id] = AgentRatingTally(agent_id, agent_type)
        return self._tallies[agent_id]

    def submit_rating(
        self,
        agent_id: str,
        agent_type: str,
        rated_by: str,
        task: str,
        score: float,
        comment: str = "",
    ) -> RatingEntry:
        """Submit one peer rating for *agent_id*."""
        tally = self.ensure_agent(agent_id, agent_type)
        entry = RatingEntry(
            agent_id=agent_id,
            rated_by=rated_by,
            task=task,
            score=score,
            comment=comment,
        )
        tally.add(entry)
        logger.debug(
            "Rating submitted: %s → %s task=%s score=%.1f",
            rated_by, agent_id, task, score,
        )
        return entry

    def submit_peer_ratings(
        self,
        performing_agent_id: str,
        performing_agent_type: str,
        task: str,
        peer_agents: List[Dict[str, str]],
        result: Any = None,
    ) -> List[RatingEntry]:
        """
        Ask each peer agent to rate the *performing* agent.

        Scoring heuristic (peers don't have real subjective opinions, so we
        use a deterministic proxy based on task name and result content):
        - Base score 75
        - +10 if result is a non-empty dict with no "error" key
        - +5  if the task succeeded (result truthy)
        - +10 if result indicates an optimised/tuned outcome
        - −20 if result contains an obvious failure indicator
        """
        entries: List[RatingEntry] = []
        base = _score_result(task, result)

        for peer in peer_agents:
            peer_id = peer.get("agent_id", "unknown")
            peer_type = peer.get("agent_type", "unknown")
            if peer_id == performing_agent_id:
                continue  # agents don't rate themselves

            # Add a small deterministic jitter per peer so scores differ
            jitter = (hash(peer_id + task) % 11) - 5  # −5 … +5
            score = max(0.0, min(100.0, base + jitter))

            comment = _build_comment(peer_type, task, score)
            entry = self.submit_rating(
                agent_id=performing_agent_id,
                agent_type=performing_agent_type,
                rated_by=peer_id,
                task=task,
                score=score,
                comment=comment,
            )
            entries.append(entry)

        return entries

    def get_tally(self, agent_id: str) -> Optional[AgentRatingTally]:
        return self._tallies.get(agent_id)

    def get_leaderboard(self) -> List[Dict[str, Any]]:
        """Return all agents sorted by average score (highest first)."""
        board = [t.to_dict() for t in self._tallies.values()]
        board.sort(key=lambda x: (-x["average_score"], -x["total_ratings"]))
        for rank, entry in enumerate(board, start=1):
            entry["rank"] = rank
        return board

    def get_agent_history(self, agent_id: str, limit: int = 20) -> List[Dict[str, Any]]:
        tally = self._tallies.get(agent_id)
        return tally.get_history(limit) if tally else []

    def get_stats(self) -> Dict[str, Any]:
        tallies = list(self._tallies.values())
        if not tallies:
            return {"total_agents": 0, "total_ratings": 0, "global_average": 0.0}
        total_ratings = sum(t.total_ratings for t in tallies)
        all_scores = [e.score for t in tallies for e in t._history]
        global_avg = round(sum(all_scores) / len(all_scores), 2) if all_scores else 0.0
        return {
            "total_agents": len(tallies),
            "total_ratings": total_ratings,
            "global_average": global_avg,
        }


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _score_result(task: str, result: Any) -> float:
    score = 75.0

    if result is None:
        return score - 15.0

    if isinstance(result, dict):
        if "error" in result or result.get("reason", "").endswith("failed"):
            return score - 20.0
        if result.get("optimised") or result.get("tuned"):
            score += 10.0
        if result.get("interference") is False:
            score += 5.0
        if isinstance(result.get("steps"), list) and len(result["steps"]) > 0:
            score += 5.0
        if "anomalies" in result and len(result["anomalies"]) == 0:
            score += 5.0
    elif result:
        score += 5.0

    return min(score, 100.0)


def _build_comment(peer_type: str, task: str, score: float) -> str:
    if score >= 90:
        return f"{peer_type}: excellent performance on {task}"
    if score >= 75:
        return f"{peer_type}: solid execution of {task}"
    if score >= 60:
        return f"{peer_type}: acceptable result for {task}"
    return f"{peer_type}: suboptimal outcome on {task}"

File: lib/admin/ratings.ts
import { v4 as uuidv4 } from "uuid"

// ─── Types ───────────────────────────────────────────────────────────────────

export interface RatingEntry {
  id: string
  agentId: string
  agentType: string
  ratedBy: string
  task: string
  score: number       // 0–100
  comment: string
  timestamp: number
}

export interface AgentTally {
  agentId: string
  agentType: string
  averageScore: number
  lastScore: number | null
  totalRatings: number
  trend: "up" | "down" | "neutral"
  rank: number
}

// ─── In-Memory Store ─────────────────────────────────────────────────────────

const MAX_HISTORY = 500

// agentId → list of rating entries (capped)
const ratingsByAgent = new Map<string, RatingEntry[]>()

// agentId → { agentId, agentType } metadata
const agentMeta = new Map<string, { agentId: string; agentType: string }>()

// ─── Core helpers ─────────────────────────────────────────────────────────────

function getOrInitAgent(agentId: string, agentType: string): RatingEntry[] {
  if (!agentMeta.has(agentId)) {
    agentMeta.set(agentId, { agentId, agentType })
  }
  if (!ratingsByAgent.has(agentId)) {
    ratingsByAgent.set(agentId, [])
  }
  return ratingsByAgent.get(agentId)!
}

function computeAverage(entries: RatingEntry[]): number {
  if (entries.length === 0) return 0
  const sum = entries.reduce((acc, e) => acc + e.score, 0)
  return Math.round((sum / entries.length) * 100) / 100
}

function computeTrend(entries: RatingEntry[]): "up" | "down" | "neutral" {
  if (entries.length < 2) return "neutral"
  const recent = entries.slice(-5).map((e) => e.score)
  const older = entries.slice(-10, -5).map((e) => e.score)
  const recentAvg = recent.reduce((a, b) => a + b, 0) / recent.length
  const olderAvg =
    older.length > 0
      ? older.reduce((a, b) => a + b, 0) / older.length
      : recentAvg
  const delta = recentAvg - olderAvg
  if (delta > 2) return "up"
  if (delta < -2) return "down"
  return "neutral"
}

// ─── Public API ───────────────────────────────────────────────────────────────

/** Submit a single peer rating. */
export function submitRating(
  agentId: string,
  agentType: string,
  ratedBy: string,
  task: string,
  score: number,
  comment: string = ""
): RatingEntry {
  if (score < 0 || score > 100) {
    throw new RangeError(`Score must be 0–100, got ${score}`)
  }
  const entries = getOrInitAgent(agentId, agentType)
  const entry: RatingEntry = {
    id: uuidv4(),
    agentId,
    agentType,
    ratedBy,
    task,
    score: Math.round(score * 100) / 100,
    comment,
    timestamp: Date.now(),
  }
  entries.push(entry)
  // Cap history
  if (entries.length > MAX_HISTORY) {
    entries.splice(0, entries.length - MAX_HISTORY)
  }
  return entry
}

/** Ensure an agent appears in the tally (even before it has any ratings). */
export function ensureAgent(agentId: string, agentType: string): void {
  getOrInitAgent(agentId, agentType)
}

/** Return the competitive leaderboard, highest score first. */
export function getLeaderboard(): AgentTally[] {
  const board: AgentTally[] = []
  for (const [agentId, entries] of ratingsByAgent.entries()) {
    const meta = agentMeta.get(agentId)!
    board.push({
      agentId,
      agentType: meta.agentType,
      averageScore: computeAverage(entries),
      lastScore: entries.length > 0 ? entries[entries.length - 1].score : null,
      totalRatings: entries.length,
      trend: computeTrend(entries),
      rank: 0,
    })
  }
  board.sort((a, b) => b.averageScore - a.averageScore || b.totalRatings - a.totalRatings)
  board.forEach((entry, i) => {
    entry.rank = i + 1
  })
  return board
}

/** Return recent rating history for one agent. */
export function getAgentHistory(agentId: string, limit = 20): RatingEntry[] {
  const entries = ratingsByAgent.get(agentId) ?? []
  return entries.slice(-limit)
}

/** Global stats summary. */
export function getRatingsStats(): {
  totalAgents: number
  totalRatings: number
  globalAverage: number
} {
  let totalRatings = 0
  let scoreSum = 0
  for (const entries of ratingsByAgent.values()) {
    totalRatings += entries.length
    scoreSum += entries.reduce((a, e) => a + e.score, 0)
  }
  return {
    totalAgents: ratingsByAgent.size,
    totalRatings,
    globalAverage: totalRatings > 0 ? Math.round((scoreSum / totalRatings) * 100) / 100 : 0,
  }
}

// ─── Seed demo data (so the leaderboard isn't empty on first load) ────────────

const AGENT_SEEDS = [
  { id: "ai-agent-seed", type: "ai_agent" },
  { id: "frequency-agent-seed", type: "frequency_agent" },
  { id: "modulation-agent-seed", type: "modulation_agent" },
  { id: "firmware-agent-seed", type: "firmware_agent" },
  { id: "comms-agent-seed", type: "comms_agent" },
]

const TASKS = ["auto_optimise", "detect_interference", "fine_tune", "adaptive_select", "recommend_config"]

let seeded = false
export function seedDemoRatings(): void {
  if (seeded) return
  seeded = true

  for (const agent of AGENT_SEEDS) {
    const peers = AGENT_SEEDS.filter((p) => p.id !== agent.id)
    // Simulate 10 rounds of peer ratings per agent
    for (let round = 0; round < 10; round++) {
      const task = TASKS[round % TASKS.length]
      for (const peer of peers) {
        const base = 60 + (Math.abs(agent.id.charCodeAt(0) * 3 + round * 7) % 35)
        const jitter = (Math.abs(peer.id.charCodeAt(0) + round) % 11) - 5
        const score = Math.max(0, Math.min(100, base + jitter))
        submitRating(agent.id, agent.type, peer.id, task, score, `${peer.type}: round ${round + 1} evaluation`)
      }
    }
  }
}
